refactor: microservice security cheat sheet (#1)
dadrus/CheatSheetSeries: cheatsheets/Microservices_Security_Cheat_Sheet.md

## Introduction

The microservice architecture is increasingly used to design and implement application systems in both cloud-based and on-premise environments, particularly for high-scale applications and services. However, it introduces a range of security challenges that must be addressed during both the design and implementation phases.

> Two of the most critical security concerns are authentication and authorization. As such, it is essential for application security architects to understand and correctly apply architectural patterns that support these concerns in microservices-based systems.

Reviewer: ```suggestion
Two of the most critical security concerns are authentication and authorization. As such, it is essential for application security architects to understand and correctly apply architectural patterns that implement these concerns in microservices-based systems.
```

Author: paragraph updated to reflect your suggestion

## Authorization Reference Architecture

To lay the foundation for the patterns described in this cheat sheet, this section introduces the general building blocks of an authorization system, based on [NIST SP 800-162](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-162.pdf). While that standard focuses on Attribute-Based Access Control (ABAC), the architectural components it defines are relevant to nearly any access control system.

![Authorization Reference Architecture](../assets/Authorization_Reference_Architecture.svg)

These functional components are:

* **Policy Administration Point (PAP):** Manages policies—offering tools for writing, testing, and updating access control logic.
* **Policy Decision Point (PDP):** Evaluates policies and computes decisions about access based on the incoming request and relevant attributes.
* **Policy Enforcement Point (PEP):** Intercepts requests and enforces the access decision provided by the PDP.

> * **Policy Information Point (PIP):** Supplies attribute data or contextual information that the PDP requires to evaluate a policy.

Reviewer: Recommendation: I'd add a short definition of Subject and Object here, too. 

Maybe also one for the repositories, though they are not necessarily different components depending on the architectural pattern (from "everything in application-db" to the picture shown above.)?

edit: It gets clear in the story, though, so maybe just leave as is. up to you

Author: Yeah, I was thinking about that as well, but decided to address those elements in the story section.

Repositories are indeed not described explicitly. Because of the name build from "context" and "repository" I thought those elements are self-descriptive enough.

But you're right, I should add a short definition for each to avoid scattering of information throughout the document.

* Alice can read up to three full articles per day on a free plan.
* Alice can read that set of articles

In these cases, the PDP returns not just a binary decision, but a decision along with **obligations, conditions, or structured attributes** that describe *how* access is permitted - such as what parts of a resource are visible, or what usage limits apply.

The PEP then takes that decision and enforces it - meaning the request is either allowed to proceed to the protected resource or is blocked. Enforcement is binary: permit or deny. If the decision includes additional data, it is up to downstream components, such as business logic or resource handlers to interpret and apply those, for example, by shaping the response or limiting available actions.

However, there is one essential prerequisite: the system must know who the subject is, that is, it must verify that Alice is indeed Alice. This is the domain of authentication. Without it, the PEP has no basis on which to enforce access decisions. Authentication is therefore foundational, which is why we begin by examining authentication patterns and approaches.

> ### First-Party vs. Third-Party Authorization

Reviewer: in the section before you talk about examining auth**entication** patternsin the last sentence, the next heading says "...Authorization". This is a bit irritating i'd say.

I'd change it to e.g. 

Headline: "First-Party context vs. Third-Party context" 

First section: "Before we look closer at authentication patterns, there's a crucial differentiation one has to be aware of: Is the context we're in a First-party, or a third-party context: ..." 

- that said, maybe it makes more sense to move it after you talked about authentication patterns. It's more relevant in the space of AuthZ, imo.

Author: While I agree with the the second part of your comment, I think, this topic is so essential, that it must be introduced at the very beginning. Although the "remainder" of the document does not make use of these definitions yet, I'll definitely want to establish corresponding links.

 There is yet an additional important topic, I don't have a name for: when a subject authorizes an action, like e.g. money transfer in banking context.

* **[Protocol-agnostic identity propagation](#protocol-agnostic-identity-propagation):** Verified identity information can be propagated to internal services using trusted, implementation-independent formats (e.g., via a newly issued [JWT](https://www.rfc-editor.org/rfc/rfc7519), injected headers carrying identity information in protected form using standards like [HTTP Message Signatures](https://www.rfc-editor.org/rfc/rfc9421.html)), or signed proprietary structures, avoiding the need to handle raw external authentication data.

#### Cons

* **Limited granularity:** Fine-grained or per-endpoint authentication policies (e.g., step-up authentication) are generally harder to implement and may require additional coordination with downstream services. This heavily depends on the capabilities of the edge proxy
* **Identity propagation challenges:** Ensuring secure and reliable propagation of identity context (e.g., via headers) requires strict validation and trust models between the edge and internal services. Proper governance can help overcome this limitation.

> * **Single Point of Failure:** ...

Reviewer: missing description

Author: Ah.. good catch! Thank you!

## Identity Propagation Patterns

> In modern service architectures, especially those aligned with Zero Trust principles, how identity is propagated internally is just as important as how it is initially authenticated. Identity propagation patterns define how identity context flows between services and determine where and how access control decisions are made.

Reviewer: you mention those principles a few times, but you never define them (or added a link to a definition). I'd do this, as it is a loaded word (as so often) and everyone would think about their own definition.

Author: Frankly speaking, I don't get your point. Do you mean

* adding links to this section in the places, which mentions the "identity propagation" principle?
* describing what Zero Trust (networking) is, respectively adding a link to e.g. NIST?
* describing what an identity context is?
* something else?

Reviewer: Second bullet, sorry, shoul've written that. It's just about a link/short definition on zero trust :)

#### Pros

* **Improved trust model:** Downstream services do not need to trust upstream service implementations, only the authorization server.
* **Cryptographically verifiable identity:** Issued tokens are signed by a trusted authorization server, offering strong integrity guarantees.
* **Scoping and audience control:** Exchanged tokens can be restricted in scope and audience, reducing the risk of token misuse.

#### Cons

* **OAuth2-specific:** Relies on OAuth2 Token Exchange, limiting its applicability to systems using that protocol family for externally visible authentication data.
* **Service-side complexity:** Application code must integrate with the authorization server, handle token exchange logic, and manage caching or retries.
* **Latency overhead:** The token exchange process introduces additional network round-trips per request flow unless aggressively optimized.

> * **Operational dependency on the AS:** Introduces runtime dependency on the authorization server's availability and scalability.

Reviewer: That runtime dependency is basically there for 98% of cases where ppl use OAuth/OIDC AS / IdP. So not really a con of this flow, but a con of a centralized "authN-and-authZ-machine" in general i'd say. A pretty much solved problem in reality, though. 

Just wanted to mention this, I'm not sure if I'd remove this con (as it is valid) - up to you :)

Author: Compared to other identity propagation patterns, this is the only one with a "hard" dependency on the IdP.  "hard" in the sense, that without the IdP being available a service cant request a new token. In all other cases, similar dependency can (but doesn't need to) exist for token verification purposes. IMO this is an important nuance.

### Protocol-Agnostic Identity Propagation

> The external request is authenticated at the system edge by a trusted component, which then generates a cryptographically signed (and/or encrypted) data structure representing the external entity’s identity and attributes (e.g., user ID, roles, permissions). This identity structure is propagated downstream to internal microservices. Internal services trust the signature from the edge issuer and use the identity structure to make access control decisions.

Reviewer: ```suggestion
The external request is authenticated at the system edge by a trusted component, which then generates a cryptographically signed (and/or encrypted) data structure representing the external subject’s identity and attributes (e.g., user ID, roles, permissions). This identity structure is propagated downstream to internal microservices. Internal services trust the signature from the edge issuer and use the identity structure to make access control decisions.
```

Author: I fear that update would break the meaning. I used entity by intention - in first party context, that would be indeed the subject, and in third-party context, it could be the user (PDP), or a client application (subject). I think it is worth referencing these two here.

Reviewer: Okay, with you here.

#### Pros

* **Cryptographic trust:** Signed tokens provide strong guarantees about the integrity and authenticity of the propagated identity.
* **Decoupling from external authentication data:** Internal services do not need to cope with protocols used at the edge or to validate externally used authentication data (such as access tokens or cookies) themselves, simplifying service logic.
* **Rich identity context:** Allows inclusion of fine-grained identity and authorization metadata.
* **Secure across trust boundaries:** Suitable for multi-tenant and zero-trust environments.

> * allows for decoupling of external entities from their internal representations, which highly enhances privacy.

Reviewer: The last bullet point falls in line with Pro #2 - **Decoupling from authentication data** - I'd add it there instead of its own bullet point.

Author: I indeed failed finding a proper name for that bullet point.

I think, there is either an issue with the link shared by you, or I don't see, what you mean.

Reviewer: meant bullet point number 2 here, "decoupling from external authentication data". I see that this is a little different though, but maybe not. Let's talk later.

Author: Actually, the last bullet point is talking about different identifiers for the externally visible id and internally used id of the entity

> #### Cons

Reviewer: I'm missing the service-side complexity con here a bit. It's still needed to validate the internal token (or whatever representation) on the service-side, and possibly without a mature ecosystem like OAuth/OIDC/JWT provide. WDYT?

Author: hm... actually, I disagree: Compared to "Simple Service-Level Identity Forwarding" a downstream service has to verify the internally used structure. But unlike the "Token Exchange-Based Identity Issuance" and "External Identity Propagation" there is no dependency on any protocol. That means the (im-)maturity of any ecosystem doesn't play a role at all. Only the ease of use of a specific structure/format (issued at the edge) and the ease of its verification plays a role. And that can even be something completely proprietary - like Netflix' Passport structure.

Reviewer: let's talk later.

Author: Resolved ;)

* **Key management complexity:** Requires secure handling and rotation of signing keys to maintain trust.
* **Token size overhead:** Signed data structures issued by the edge component may be large, increasing network overhead.
* **Revocation challenges:** Once issued, signed data structures may be valid for many services until expiration, complicating immediate revocation. That can however be mitigated by issuing short living signatures and by creating downstream service specific structures.

> * **Increased complexity at the edge:** The edge component must handle token signing and may become a critical security point.

Reviewer: I'd say it may become "a critical security component", point sounds kinda weird. 

Now that said, I think that it'd also be a critical component nevertheless, so maybe rephrase this like so: 

**Increased complexity at the edge:** The edge component must handle external token verification, internal token generation / signing and as such becomes a critical security component.

* **Built-in encryption and authenticity:** mTLS secures all communication at the transport layer.
* **Protocol-agnostic:** Works transparently for HTTP, gRPC, or other protocols without changes to application logic.

#### Cons

* **Operational complexity:** Certificate issuance, rotation, and revocation require automation and infrastructure (e.g., mesh, PKI, SPIRE).
* **Limited application-level context:** Certificates provide service-level identity but lack granular attributes (e.g., purpose, scopes, tenancy) for fine-grained authorization, auditing, or delegation, requiring additional application-layer mechanisms.

### Token Based Authentication

In this pattern, the calling service (caller) authenticates itself by attaching a token to each request to another microservice (callee). The token is issued by a special security token service after the service authenticates using its credentials (e.g., service ID and a secret). Upon reception of the token, the callee can verify it (online or offline), extract the caller’s identity and further attributes and use the information for further processing or the request.

#### Pros

> * **Rich application-level context:** Tokens can carry detailed attributes (e.g., roles, tenancy, permissions), enabling fine-grained authorization and business logic at the application layer, as well as tracking of business-level actors or intents, facilitating compliance and delegated authorization.

Reviewer: just a note: for sure it is fine-grained authorization compared to mTLS, but not generally fine-grained from the view of ABAC or ReBAC-Systems. Same problems as with all token based auth apply: it's header/metadata and shouldn't be more than 8k overall. I've seen environments where the sheer amount of roles (sic!) exceeded that size and led to 413 Errors (or, more subtly, 400 or 502 errors bc the server implementation for those cases was bad). But as said: Just a note - this is usually more the case when ppl try to do "really" fine grained authZ with tokens. :)

Author: Fully agree. However, that section focuses on service-to-service authentication and context propagation. Whether the related attributes should be used for authorization purposes depends on which authorization patterns (which I discuss later) best align with the given requirements.

Based on your comment, I believe we need to clarify the role of tokens. As you know, there are various types and formats of tokens. Regardless of type or format, a token always carries attributes. It doesn't need to be self-descriptive to do so. I’m concerned that this may not be clear enough, which could easily lead to misunderstandings and confusions.

![Decentralized Service-Level Access Control](../assets/Decentralized_Service_Level_Access_Control.svg)

The access control rules are typically implemented using native language constructs (e.g., `if`/`else` statements), either inline with business logic functions or via abstraction mechanisms such as interceptors.

When a microservice receives a request containing authorization metadata (e.g., end-user context or resource identifiers), it evaluates whether access should be granted. This may involve querying other services (PIPs) for additional attributes before reaching a decision and enforcing it (implicitly). Alternatively, some services may use asynchronous communication patterns (e.g., periodic syncs or event-driven updates) to pre-fetch required data in advance, improving performance and resilience.

When adopting this approach, the following trade-offs should be considered:

> #### Pros

Reviewer: another Pro maybe missing: 
**Performance**: Given the microservice has everything at its hands to make and enforce an access decision without reaching out to services other than it's own database, for some cases - especially when dealing with access decisions for a huge set of objects - it is the most performant pattern. 

**edit**: I re-read the pro section, it's already there, just a bit shorter.

* **Familiar development model**: Developers can use the same language and tools they already know.
* **Framework support**: Many libraries and frameworks exist for many languages to reduce boilerplate and simplify integration.
* **Rapid prototyping**: Policy logic is implemented directly in code, enabling quick experimentation and iteration.

> * **Team autonomy**: Fits well with domain-driven design and independent team ownership; each team can choose its approach.

Reviewer: ...until someone maps the "access" domain and sees that this domain spans more than one service, as access boundaries and domain boundaries are not necessarily the same thing. And the business rules (who has access to what etc) and constraints (before the fact audit etc) for the "access" domain show that it was a really bad idea to do this ;) 

What I want to say here is: "Domain Driven Design" and access control is a whole story in itself, i wouldn't mention this term here.

Author: From a DDD perspective, mapping something like "access" to a domain would be considered an anti-pattern. DDD is about modeling business domains, not cross-cutting concerns like access control ;). But I see your point.

* **High performance**: Policy evaluation is done in-memory within the microservice.
* **Full context awareness**: The service has access to runtime data, business logic, and domain models, enabling fine-grained, context-rich and nuanced decisions.
* **Failure isolation**: If all required attributes are available locally or cached, failures in external systems do not impact decision-making.

> #### Cons

Reviewer: (possibly) missing Con:

**Intermingling of business- and access-related code**: Access requirements and service requirements may have distinct lifecycles. Adding authorization-related code directly to the services main codebase may result in the need to refactor and redeploy the service when there are access-related changes to be made (e.g., a subscription-based service needs a bigger rewrite because now the "silver" subscription covers less than before, or a new "platin" subscription is added)

if that makes sense. You kinda cover it in the scattered logic-part, though

Author: That’s exactly the point I intended to capture in the "scattered logic" section. I agree, it would be worthwhile to rename and rephrase that con to better reflect this, and to incorporate your suggestion as well.

* **Scattered logic**: Authorization requirements tend to spread across multiple microservices, leading to code duplication, increased complexity, and maintenance overhead. Over time, this results in a slow and error-prone policy lifecycle, significantly reducing time to market. This is a classic “Hardcoded Rules” antipattern.
* **Role explosion**: Business stakeholders typically describe authorization requirements using roles - for example, “a user with role X can do Y.” Without introducing an abstraction layer between business roles and the actual implementation, systems often accumulate many similar but inconsistent roles. Roles also tend to evolve or change names over time. This leads quickly to role explosion, again slowing the policy lifecycle and increasing the risk of errors. This is known as the “Code Against the Role” antipattern.

> * Inconsistent interpretations: Autonomous teams may interpret and implement policies differently, making consistent governance across the system nearly impossible. This results in enforcement gaps and unpredictable behavior.

Reviewer: I'd rewrite it like this:

```suggestion
* **Deprived Governance**: Autonomous teams may interpret and implement policies differently, making consistent governance for the whole environment nearly impossible. This may result in enforcement gaps and unpredictable behavior.
```

* **No central auditability**: When authorization logic is distributed across services, it becomes nearly impossible to answer "before-the-fact" questions such as "Who has access to what, and when?" - a key requirement in compliance and security contexts.
* **Inconsistent monitoring**: Logging and audit trails vary widely across services and are often incomplete or incompatible. This hampers the ability to detect abuse, investigate incidents, or analyze system-wide access patterns.
* **Coverage gaps**: Many frameworks do not expose ways to integrate access control into certain auto-exposed endpoints. Teams may also forget to secure these paths entirely. Documentation of the frameworks is also often inconsistent or misleading. All of that leads to unintended public exposure of sensitive endpoints.

These cons often result in "accept by default" behavior, ultimately leading to broken access control vulnerabilities.

### Centralized Service-Level Access Control with embedded PDP

> This pattern aims to address the first three drawbacks of the previous pattern, namely the "Hardcoded Rules" and "Code Against the Role" antipatterns. The goal is to reduce complexity, improve time to market, and establish governance over policy definitions. In this model, authorization rules are defined independently from the microservice code. This separation allows policies to be reviewed, versioned, and audited without being tied to the specific implementation languages of the microservices. These policies can reside in a dedicated policy repository, which explains the "centralized" in the pattern name, or they can be colocated with the service code in the same repository. The essential aspect is that policies are decoupled from the service code, rather than centralized in infrastructure terms. The actual evaluation of access decisions still takes place locally to each microservice using an embedded PDP.

Reviewer: ```suggestion
This pattern aims to address the first two drawbacks of the previous pattern, namely the "Hardcoded Rules" and "Code Against the Role" antipatterns. The goal is to reduce complexity, improve time to market, and establish governance over policy definitions. In this model, authorization rules are defined independently from the microservice code. This separation allows policies to be reviewed, versioned, and audited without being tied to the specific implementation languages of the microservices. These policies can reside in a dedicated policy repository, which explains the "centralized" in the pattern name, or they can be colocated with the service code in the same repository. The essential aspect is that policies are decoupled from the service code, rather than centralized in infrastructure terms. The actual evaluation of access decisions still takes place locally to each microservice using an embedded PDP.
```

Reviewer: alternatively, also name the third ;)

Reviewer: another thing about this paragraph:
> The essential aspect is that policies are decoupled from the service code, rather than centralized in infrastructure terms. 

This sentence doesn't really make sense for me. the "rather..." part, especially. it is essential that service code != policy code, 100% agree. But it might be perfectly possible that this pattern is used and policies would still be written by a centralized infra team. In fact, I understand the "centralized" part as in the old owasp sheet, so that "[...] access control rules are defined centrally but stored and evaluated at the microservice level". The rules are defined centrally, by whatever team that takes.

Reviewer: oh... re-reading, it says "terms"... nvm. I still found it a bit confusing, i'd rewrite it like this: 

> The essential aspect is that policies are decoupled from the service code. Access control rules are defined centrally (e.g. in the "policy repository") and get stored and evaluated at the microservice by the embedded PDP and PEP.

Author: > alternatively, also name the third ;)

Yeah, the sentence is a bit weird. But the next sentence clarifies that ;):  "The goal is to reduce complexity, improve time to market, and establish governance over policy definitions" - here all three are mentioned :)

Author: > oh... re-reading, it says "terms"... nvm. I still found it a bit confusing, i'd rewrite it like this:
> 
> > The essential aspect is that policies are decoupled from the service code. Access control rules are defined centrally (e.g. in the "policy repository") and get stored and evaluated at the microservice by the embedded PDP and PEP.

Hmm… I’m not quite comfortable with that phrasing, as it implies that access control rules must be stored centrally, which isn’t necessarily the case. Rules can still live alongside the code in the same repository. It really depends on who owns a particular rule and how rules are composed. Having a central repository for globally relevant rules, maintained by teams like infrastructure or security, is perfectly valid, just as it is to keep service-specific policies within the service’s own repo. Rephrasing the original description risks losing that important distinction - and that, in turn, can lead to misconceptions and misunderstandings.

Reviewer: hmm, you are correct I think. Have to re-read it again. Let's talk later.

Author: ... The essential aspect is that policies are decoupled from the service code, rather than intertwined with it.

> The PDP can be implemented as a library (e.g., [Casbin](https://casbin.org/)) embedded in the service’s codebase, or as a local sidecar process (e.g., [Open Policy Agent](https://www.openpolicyagent.org/)). Authorization rules are now defined using the PDP’s domain-specific language (e.g., Rego in the case of OPA), rather than being hardcoded into the service logic. Typically, the PDPs with this pattern implement the so-called Policy-Based Access Control (PBAC) approach.

Reviewer: I'd add a little explainer that PBAC is essentially an implementation of ABAC.

Author: On one hand, yes - that was my intention as well. On the other hand, I’m a bit hesitant to include that description, as it might open up a lot of additional questions that I’m not keen on addressing as part of this cheat sheet update. You know, what I'm talking about ;)

Reviewer: correct. :) one has to draw a line somewhere.

* **Good performance:** Authorization decisions are computed locally, either in-memory (for library-based PDPs) or by a co-located sidecar PDP (communicated with over localhost), resulting in negligible latency.
* **Improved monitoring:** All decisions can be consistently logged and monitored, assuming proper instrumentation.
* **Team autonomy:** Teams remain responsible for their services and their policies, with local enforcement and minimal external dependencies. This aligns well with independent team ownership and domain-driven design principles.
* **Failure isolation:** Services remain resilient as long as required decision attributes are locally available. No central dependency for decision evaluation.
* **Enhanced testability:** Authorization logic can be tested independently of the microservice business logic.

#### Cons

* **Policy distribution complexity:** Policies are now decoupled from the code, so mechanisms are needed to deploy the correct version of each policy to the appropriate service instances.

> * **Context sharing:** PDPs do not inherently have access to the microservice context. Developers must design mechanisms to assemble and pass the right attributes into the PDP for evaluation. Even though efforts have been put to address that (e.g. the [AuthZEN Authorization API](https://openid.net/authzen-authorization-api-1-0-implementers-draft-approved/)), there are many more topics to be addressed in addition.

Reviewer: ```suggestion
* **Context sharing:** PDPs do not inherently have access to the microservice context. Developers must design mechanisms to assemble and pass the right attributes into the PDP for evaluation. There are ongoing early-stage efforts to address these needs (e.g. the [OpenID AuthZEN Initiative](https://openid.net/authzen-authorization-api-1-0-implementers-draft-approved/)).
```

Reviewer: A bit of renaming, and I removed the part that "there are many more topics to be addressed in addition" and added "early-stage" instead, because otherwise I'd expect the topics to be listed, and it might sound offensive (which is not intended for sure ;) ).

Author: Remove the second sentence and reference to the data distribution strategies

Although this pattern improves observability and supports a broader range of access control models, it introduces its own trade-offs and does not eliminate all challenges found in the previous pattern.

#### Pros

* **Policy governance:** Policies can be centrally defined, versioned, reviewed, and audited, independent of the service’s implementation language.
* **Policy layering:** The model allows for both global (e.g. security team–defined) and local (e.g. service team–defined) policies to coexist. This enables clearer separation of concerns and better alignment with organizational structure and responsibilities.
* **Improved monitoring:** All decisions can be consistently logged and monitored, assuming proper instrumentation.
* **Team autonomy:** Teams remain responsible for their services and their integration with the PDP. This aligns well with independent team ownership and domain-driven design principles.

> * **Support for "before-the-fact" audit:** Particularly with ReBAC and NGAC systems as PDP, authorization models allow querying the existing access rights making answering the corresponding questions a simple game.

Reviewer: ```suggestion
* **Support for "before-the-fact" audit:** Particularly with ReBAC and NGAC systems as PDP, authorization models allow querying the existing access rights, making answering the corresponding questions a simple game.
```

Reviewer: not entirely sure about the comma. no native speaker. but my intuition says there needs to be one.

* **Support for additional PDP models:** This pattern enables the use of a broader set of access control models, including ReBAC and NGAC systems, which may better fit for a given business context.

#### Cons

* **Policy distribution complexity:** Policies are managed independently of service code, and the PDP may support multiple services and domains. Ensuring the correct policy version is applied consistently can be complex.
* **Context sharing:** PDPs do not inherently have access to service context. Mechanisms must be designed to assemble and pass the right attributes into the PDP for evaluation and in case of the ReBAC/NGAC systems to build the actual authorization model.

> * **Performance overhead:** Network hops between the microservice and the external PDP introduce latency and a potential point of contention under high load.

Reviewer: just a digression, nothing to change: at least spiceDB is also usable as a library -> https://github.com/authzed/examples/tree/main/spicedb-as-library - but it's "just" a community example (and not maintained i guess, still on v1.25), and iirc experimental, so not really worth mentioning i guess, also bc imo rebac systems (or at least spiceDB) are really "designed" to work as their own, highly available and horizontally distributable service.

It might come in handy when you want to build your own abstraction around SpiceDB, though, or enhance it with a wrapper for more fine-grained meta-access-control currently only available in the paid version, etc... but as said: I digress. Nevermind ;)

Author: Yeah. But we would be talking about embedded PDP then ;)

* **Incomplete enforcement observability:** While policy decisions are consistently logged by the PDP, there’s still limited visibility into whether those decisions were enforced correctly in the service code. Instrumentation gaps can hinder detection of abuse, validation of protections, and system-wide access analysis.

> * **Coverage gaps:** There is no guarantee that every service or endpoint consistently integrates with the central PDP. Some frameworks expose endpoints by default, often without offering hooks for policy enforcement. Teams may also just forget to add the required logic to some endpoints. Combined with poor or misleading documentation, this can result in unintentionally exposed functionality and missed access control.

Reviewer: > Some frameworks expose endpoints by default, often without offering hooks for policy enforcement.

I stumble about this sentence now the second time, and it's not 100% clear for me what you mean here. Can you explain? You men stuff like springs [actuator-endpoints](https://docs.spring.io/spring-boot/reference/actuator/endpoints.html#actuator.endpoints.controlling-access), that are not OOTB-usable with an external PDP?

Author: E.g. Spring actuator endpoints, fastapi (python) exposed endpoints for the same purposes, or the openapi related endpoints. Static routes e.g. in express.js, etc

Reviewer: Okay, understood. Might be worth to add some examples ;)

As with the previous patterns, some gaps remain - particularly around enforcement coverage and observability - which can result in “accept by default” behaviors and ultimately lead to broken access control vulnerabilities.

### Edge-Level Authorization (Classic)

This pattern aims to address several shortcomings of Decentralized Service-Level Access Control, particularly inconsistent enforcement, policy sprawl, and limited observability. Instead of embedding authorization logic within each service, access control is moved to the system’s perimeter—typically implemented via API gateways, ingress controllers, or reverse proxies.

![Edge-Level Authorization (Classic)](../assets/Edge_Level_Authorization_Classic.svg)

Since authorization must follow authentication, this pattern tightly couples authentication and authorization at the network boundary. Gateways or proxies serve as the Policy Enforcement Point (PEP), and either evaluate policies locally (using embedded logic or libraries, similar to the “centralized with embedded PDP” pattern) or delegate decisions to an external Policy Decision Point (PDP) (as in the “centralized service-level access control” pattern).

> Since all external traffic flows through the edge component, this is the first pattern that guarantees every inbound request is observed and subject to access control logic. This minimizes the risk of unnoticed “accept by default” behavior and establishes a consistent enforcement point for all inbound traffic—but introduces its own set of trade-offs.

Reviewer: > This minimizes the risk of unnoticed “accept by default” behavior ...

I'd challenge that. It just "shifts the problem to one central place". It is still perfectly possible to forget to add the necessary policy, and with a  default that is not whitelisting but blacklisting, it would still open up broken access control vulns.

Author: I didn’t claim it’s a silver bullet. The key points are:

> ... This **minimizes the risk** of unnoticed “accept by default”

> ... ensuring uniform application of policies and **reducing the likelihood** of unprotected endpoints

Broken access control can still occur. But it’s significantly less likely compared to other patterns. Of course, this depends heavily on the tools and configurations used.

How would you suggest rephrasing that sentence to include a bit of a heads-up?

Reviewer: Let's talk later.

Author: Remove the sentence "This minimizes the risk of unnoticed “accept by default” behavior and establishes a consistent enforcement point for all inbound traffic—but introduces its own set of trade-offs." - it is already part of the pros

#### Pros

* **Consistent enforcement:** All inbound requests pass through a centralized enforcement point, ensuring uniform application of policies and reducing the likelihood of unprotected endpoints ("no accept by default").
* **Policy governance:** Policies can be centrally defined, versioned, reviewed, and audited, independent of the service’s implementation language.

> * **Best observability:** All external access attempts are visible and can be logged centrally, supporting effective monitoring, alerting, and forensics.

Reviewer: Depending on the networking configuration, also internal (service-2-service) access attempts are visible - though these are more often than not "accept by default" (o hai k8s). But they can (o hai cilium).

Reviewer: okay, I just read the con for this point, all good1.

> * **Support for "before-the-fact" audit:** Particularly with ReBAC and NGAC systems as PDP, authorization models allow querying the existing access rights making answering the corresponding questions a simple game.

Reviewer: comma may be missing, same as above. Feel free to leave as is ;)

* **Rapid prototyping:** Through authorization contracts, teams can experiment with different authorization models (e.g., embedded JWT claims, header-based roles, etc.) without relying on the infrastructure components.
* **Fine-grained context:** The proxy can fetch contextual data from arbitrary PIPs, enabling context-sensitive decisions based on domain-specific attributes, object metadata, or user state.
* **Service autonomy:** Authorization contracts empower microservice teams to define their own access control needs declaratively, supporting domain-driven service ownership without duplicating enforcement logic.
* **Protocol-agnostic identity propagation:** The system can rewrite identity and authorization data into formats that match each service’s expectations (e.g., structured JWTs, plain or signed headers), decoupling service specific logic from authentication or authorization protocols.
* **Secure by Default:** The use of declarative contracts and centralized enforcement reduces misconfiguration risks and prevents implicit access grants.

> #### Cons

Reviewer: I'd still miss the classic con of 

> Enforcement blind spots and defense-in-depth violations: Since the edge only governs ingress traffic, any internal traffic (e.g., service-to-service calls) or network misconfigurations may bypass enforcement entirely - violating the defense-in-depth principle and creating a single point of failure.

at least without more explanation, it is not clear why it shouldn't apply here, too.

Author: actually, because of protocol agnostic identity propagation and because of authorization contracts in use.

Reviewer: yes, I as myself know this. But as a reader without that knowledge, it just feels alien without a good example of these, why it shouldn't apply here. Let's talk :)

Author: I'll add an example to showcase that

**Change Rate**

This dimension describes how frequently a policy is expected to change, which has implications for where and how policies should be reviewed, deployed, and versioned.

* **High:** Frequently changing policies (days to weeks) requiring agile authoring processes, often at the microservice or domain level. Example: A marketing service adjusts promotional offer criteria weekly based on campaign feedback.
* **Medium:** Policies changing monthly or quarterly, typically at the microservice or domain level. Example: A billing service updates discount policies each quarter based on market trends.
* **Low:** Stable policies with formal review, typically at the domain or central level. Example: GDPR-driven data access policies updated annually or less frequently.

### Policy Distribution Strategies

Distributing policies to PDPs ensures they are available for evaluation in microservice architectures. This subsection outlines two primary strategies, pre-loaded policies and embedded policies, each with trade-offs affecting performance, scalability, and policy freshness. The choice mainly depends on the Policy Characteristics.

#### Pre-Loaded Policies

> Policies are proactively sent to the PDP and stored locally for evaluation, often alongside pre-loaded data, as described in Pre-Loaded Data.

Reviewer: please add a reference / internal link to #pre-loaded-data
